C/cpu.c

```diff
@@ -0,0 +1,191 @@
+  m2 = _mm_sha256msg2_epu32(_mm_add_epi32(m2, _mm_alignr_epi8(m1, m0, 4)), m1);
+  m0 = _mm_sha256msg1_epu32(m0, m1);
+
+  /* Rounds 40 - 43 */
+  msg = _mm_add_epi32(m2, _mm_set_epi64x(0xC76C51A3C24B8B70ULL, 0xA81A664BA2BFE8A1ULL));
+  s1 = _mm_sha256rnds2_epu32(s1, s0, msg);
+  s0 = _mm_sha256rnds2_epu32(s0, s1, _mm_shuffle_epi32(msg, 0x0e));
+  m3 = _mm_sha256msg2_epu32(_mm_add_epi32(m3, _mm_alignr_epi8(m2, m1, 4)), m2);
+  m1 = _mm_sha256msg1_epu32(m1, m2);
+
+  /* Rounds 44 - 47 */
+  msg = _mm_add_epi32(m3, _mm_set_epi64x(0x106AA070F40E3585ULL, 0xD6990624D192E819ULL));
+  s1 = _mm_sha256rnds2_epu32(s1, s0, msg);
+  s0 = _mm_sha256rnds2_epu32(s0, s1, _mm_shuffle_epi32(msg, 0x0e));
+  m0 = _mm_sha256msg2_epu32(_mm_add_epi32(m0, _mm_alignr_epi8(m3, m2, 4)), m3);
+  m2 = _mm_sha256msg1_epu32(m2, m3);
+
+  /* Rounds 48 - 51 */
+  msg = _mm_add_epi32(m0, _mm_set_epi64x(0x34B0BCB52748774CULL, 0x1E376C0819A4C116ULL));
+  s1 = _mm_sha256rnds2_epu32(s1, s0, msg);
+  s0 = _mm_sha256rnds2_epu32(s0, s1, _mm_shuffle_epi32(msg, 0x0e));
+  m1 = _mm_sha256msg2_epu32(_mm_add_epi32(m1, _mm_alignr_epi8(m0, m3, 4)), m0);
+  m3 = _mm_sha256msg1_epu32(m3, m0);
+
+  /* Rounds 52 - 55 */
+  msg = _mm_add_epi32(m1, _mm_set_epi64x(0x682E6FF35B9CCA4FULL, 0x4ED8AA4A391C0CB3ULL));
+  s1 = _mm_sha256rnds2_epu32(s1, s0, msg);
+  s0 = _mm_sha256rnds2_epu32(s0, s1, _mm_shuffle_epi32(msg, 0x0e));
+  m2 = _mm_sha256msg2_epu32(_mm_add_epi32(m2, _mm_alignr_epi8(m1, m0, 4)), m1);
+
+  /* Rounds 56 - 59 */
+  msg = _mm_add_epi32(m2, _mm_set_epi64x(0x8CC7020884C87814ULL, 0x78A5636F748F82EEULL));
+  s1 = _mm_sha256rnds2_epu32(s1, s0, msg);
+  s0 = _mm_sha256rnds2_epu32(s0, s1, _mm_shuffle_epi32(msg, 0x0e));
+  m3 = _mm_sha256msg2_epu32(_mm_add_epi32(m3, _mm_alignr_epi8(m2, m1, 4)), m2);
+
+  /* Rounds 60 - 63 */
+  msg = _mm_add_epi32(m3, _mm_set_epi64x(0xC67178F2BEF9A3F7ULL, 0xA4506CEB90BEFFFAULL));
+  s1 = _mm_sha256rnds2_epu32(s1, s0, msg);
+  s0 = _mm_sha256rnds2_epu32(s0, s1, _mm_shuffle_epi32(msg, 0x0e));
+
+  /* Combine with old state */
+  s0 = _mm_add_epi32(s0, so0);
+  s1 = _mm_add_epi32(s1, so1);
+
+  /* Store state */
+  t1 = _mm_shuffle_epi32(s0, 0x1B);
+  t2 = _mm_shuffle_epi32(s1, 0xB1);
+  s0 = _mm_blend_epi16(t1, t2, 0xF0);
+  s1 = _mm_alignr_epi8(t2, t1, 0x08);
+  _mm_storeu_si128((__m128i*)(midstate + 0), s0);
+  _mm_storeu_si128((__m128i*)(midstate + 4), s1);
+}
+
+#endif
+
+/* WARNING: This function isn't threadsafe and must not be called concurrently with any function in this library, including itself.
+ * It is encouranged that you call this function once in the main function of your program
+ *
+ * This function inspects the CPU's capabilities and enables the use of some optimized instructions, if they are known and available.
+ *
+ * If the CPU supports Intel sha_ni operations, the result value has the SIMPLICITY_HAS_SHA_NI_FLAG bit set.
+ */
+uint32_t simplicity_cpu_optimize_not_thread_safe(void) {
```

Comment:
AFAICS, SHA NI is not enabled by default if it is available. One has to call the `simplicity_cpu_optimize_not_thread_safe` function to explicitly opt into the optimization.